File: among_them/guided_bot/perception/interstitial.nim
## Interstitial detection — the cheap first gate of the per-frame
## pipeline. Phase 1.0.
##
## Among Them's voting screen, role-reveal, vote-result, and game-over
## screens are all rendered with a black background covering most of
## the framebuffer. Gameplay never does; off-map pixels in gameplay
## are filled with `MapVoidColor` (palette index 12), not black (see
## `bitworld/among_them/players/how_to_make_a_bot.md` § "Interstitial
## Detection"). That lets a 30 %-or-more-black-pixels test cleanly
## separate the two without false positives from off-map padding.
##
## Phase 1.0 distinguishes only "black-enough to be an interstitial"
## from "gameplay". Phase 1.5 (OCR) reads the actual interstitial
## text and classifies into role-reveal / voting / vote-result /
## game-over subtypes. Until then all interstitials are reported as
## `InterstitialUnknown`.

import ../constants
import ../types
import frame

const
  ## Minimum fraction of black pixels for a frame to be classified as
  ## an interstitial, expressed as a count threshold over `FrameLen`.
  ## Matches modulabot's `INTERSTITIAL_BLACK_PERCENT = 30` in both the
  ## Python port (`frame.py`) and the original Nim implementation
  ## (`bitworld/among_them/players/modulabot/frame.nim`). Keep this
  ## constant in lockstep.
  InterstitialBlackPercent* = 30
  InterstitialBlackThreshold* = (InterstitialBlackPercent * FrameLen + 99) div 100

type
  InterstitialObservation* = object
    ## What `detectInterstitial` reports. Embedded directly into
    ## `Percept` (see `perception.nim`) so belief update can merge it
    ## without re-running the scan.
    isInterstitial*: bool
    kind*: InterstitialKind
    blackPixelCount*: int

proc detectInterstitial*(frame: openArray[uint8]): InterstitialObservation =
  ## Classify a single unpacked frame as interstitial or gameplay.
  ##
  ## Semantics: pure black-pixel count threshold. We do NOT look at
  ## the four screen corners (that's the pre-2024 modulabot heuristic
  ## that `how_to_make_a_bot.md` documents as having broken when chat
  ## text or UI marks touched corners).
  ##
  ## Phase 1.0 returns `InterstitialUnknown` for any black-enough
  ## frame. Phase 1.5 refines that to the correct `InterstitialKind`
  ## variant via OCR.
  result.blackPixelCount = frame.blackPixelCount()
  if result.blackPixelCount >= InterstitialBlackThreshold:
    result.isInterstitial = true
    result.kind = InterstitialUnknown
  else:
    result.isInterstitial = false
    result.kind = NotInterstitial

proc phaseFromInterstitial*(
    prevPhase: GamePhase,
    obs: InterstitialObservation): GamePhase =
  ## Phase transition rule for phase 1.0. Only two states the
  ## observation can confirm with certainty: "interstitial" (some
  ## kind of gap screen) and "gameplay" (map is visible).
  ##
  ## Phase 1.5's OCR-refined interstitial kind lets us disambiguate
  ## voting vs. role-reveal vs. game-over; until then we keep any
  ## non-Unknown phase the caller had before if the interstitial
  ## observation is consistent with it, to avoid thrash.
  if obs.isInterstitial:
    case prevPhase
    of PhaseVoting, PhaseGameOver:
      prevPhase
    else:
      PhaseInterstitial
  else:
    PhaseGameplay

File: archive/players/among_them/coborg/perception/interstitial.py
"""Interstitial detection. Port of
``users/james/personal_cogs/among_them/guided_bot/perception/interstitial.nim``.

Among Them's voting / role-reveal / vote-result / game-over screens
all paint a black background over most of the frame. Gameplay never
does — off-map pixels in gameplay are filled with ``MAP_VOID_COLOR``
(palette 12), not palette 0 black. A 30%-or-more-black-pixels test
cleanly separates the two without false positives from off-map
padding.

This module is intentionally a single-shot classifier: it says
"interstitial" or "gameplay" and reports the black-pixel count.
**It does not distinguish among interstitial subtypes** (role
reveal vs voting vs vote result vs game over). That refinement
comes from OCR in S4.5 and gets layered on top of this gate.

The companion ``phase_from_interstitial`` proc in the upstream Nim
lives in the belief layer; it isn't ported here because our belief
layer is P2 work, not perception.
"""

from __future__ import annotations

from dataclasses import dataclass
from enum import Enum

import numpy as np

from .frame import FRAME_LEN, black_pixel_count

# Threshold expressed as both a percentage (for documentation) and a
# pre-computed pixel count (for the actual comparison). The Nim source
# uses ceiling division so a fixture exactly at the boundary stays
# within budget; we mirror that.
INTERSTITIAL_BLACK_PERCENT: int = 30
INTERSTITIAL_BLACK_THRESHOLD: int = (INTERSTITIAL_BLACK_PERCENT * FRAME_LEN + 99) // 100


class InterstitialKind(Enum):
    """Mirrors upstream ``InterstitialKind`` in
    ``guided_bot/types.nim``. String values are the lowercase form
    the Nim oracle emits so sidecars are human-readable.

    :data:`NOT_INTERSTITIAL` and :data:`UNKNOWN` are the only two
    values :func:`detect_interstitial` ever returns — the rest are
    subtype refinements that land with OCR in S4.5+ (the belief layer
    upgrades :data:`UNKNOWN` to the right subtype once OCR reads the
    on-screen text)."""

    NOT_INTERSTITIAL = "not_interstitial"
    UNKNOWN = "unknown"
    ROLE_REVEAL = "role_reveal"
    ROLE_REVEAL_CREWMATE = "role_reveal_crewmate"
    ROLE_REVEAL_IMPOSTER = "role_reveal_imposter"
    VOTING = "voting"
    VOTE_RESULT = "vote_result"
    GAME_OVER = "game_over"


@dataclass
class InterstitialObservation:
    """Result of one :func:`detect_interstitial` call. Mirrors
    upstream ``InterstitialObservation``."""

    is_interstitial: bool
    kind: InterstitialKind
    black_pixel_count: int


def detect_interstitial(frame: np.ndarray) -> InterstitialObservation:
    """Classify a single unpacked frame as interstitial or gameplay.

    Pure black-pixel count threshold — no four-corner heuristic, no
    structural pattern matching. The 30% threshold is documented in
    ``bitworld/among_them/players/how_to_make_a_bot.md`` § "Interstitial
    Detection" as the post-2024 replacement for the older corner check
    that broke when chat text touched a corner.

    Returns :class:`InterstitialKind.UNKNOWN` for any black-enough
    frame; OCR (S4.5) refines that to the specific subtype.
    """
    count = black_pixel_count(frame)
    if count >= INTERSTITIAL_BLACK_THRESHOLD:
        return InterstitialObservation(
            is_interstitial=True,
            kind=InterstitialKind.UNKNOWN,
            black_pixel_count=count,
        )
    return InterstitialObservation(
        is_interstitial=False,
        kind=InterstitialKind.NOT_INTERSTITIAL,
        black_pixel_count=count,
    )
